proposals/0248-tpu-feedback.md

```diff
@@ -0,0 +1,186 @@
+receive feedback from the server must be ready to receive application datagrams
+from its connections.
+
+A client connection notifies the server of its interest to receive feedback by
+sending a datagram soliciting feedback. This helps server to avoid sending
+datagrams to clients which are not interested in receiving feedback or clients
+which are not upgraded to have such capabilities.
+
+The client and server must enable the support of datagrams QUIC frame types
+via transport parameter `max_datagram_frame_size`.
+
+The server sends the feedback datagrams to interested clients periodically.
+The frequency is decided by the server and consideration should be given to
+coalesce feedback on more transactions to reduce datagram counts and short
+enough feedback latency. For example 100ms can be considered. If there are
+more transactions to be acknowledged in one datagram, more datagrams
+can be sent without waiting for the feedback interval.
+
+### The format of the TPU feedback
+
+struct TpuFeedback {
+  version: u8,
+  timestamp: u64,
+  transaction_state: TransactionState,
+  priority_fee_info: PriorityFeeInfo
+}
+
+The version is a 8 bit unsinged integer, it is set to 1 for first version. It
+is designed to allow the protocol to extend for future extension.
+
+The timestamp is the 64 bit unsigned integer representing the number of seconds
+since midnight 01/01/1970.
+
+TransactionState is defined as the following:
+
+struct TransactionState {
+  transactions_count: u8
+  (transaction, state) pairs
+}
+
+The transactions_count is 8 bit unsgined integer representing the count of
+transactions in this feedback. It is followed by the count of
+(transaction, state) pairs. Each transaction is represented by the first 8 bytes
+of the transaction signature and each state is represented by an 32 bit unsigned
+integer.
+
+The transaction state value is defined as the below enum:
+
+enum TransactionStateValue {
+  Packed = 0,
+  FeeTooLow = 1,
+  FeePayerBalanceTooLow = 2,
+  NotPackedDueToUnspecifiedReason = 2^32 -1,
+}
+
+Newer state value can be defined as needed in the future with its own unique
+value. Client should disregard the value if it does not understand its meaning.
+
+The proirity fee information is defined as the following,
+
+struct PriorityFeeInfo {
+  priority_fee_stats: {
+    min_priority_fee/compute unit ratio
+    median_priority_fee/compute unit ratio
+    max_priority_fee/compute unit ratio
+  }
+}
+
+These ratio are represented as f64 numbers using 8 bytes.
+
+## Impact
+
+There will be higher cost for the server to send the feedback to all connected
+clients in term of network bandwidth and compute resources. With expected
+2500 connections, and the nax 1232 bytes per, and assuming packets are sent
+per 100ms the network cost is:
+
+2500*1232*10 bytes /s ~= 30 MB/s
+
+For sending only priority info to idle connections, the size of the feedback
+datagram is smaller -- 33 bytes.
+
+The cost is:
+
+2500*21*10 bytes /s = 0.8 MB
```

Comment:
you are not counting the headers. IP + UDP+QUIC are already over 30 bytes, so you get some 12 Mbps.